electron/js/menu/TrayIconHandler.js

```diff
@@ -0,0 +1,85 @@
+const {app, Menu, nativeImage, Tray} = require('electron');
+const config = require('./../config');
+const environment = require('./../environment');
+const lifecycle = require('./../lifecycle');
+const locale = require('./../../locale/locale');
+const path = require('path');
+const windowManager = require('./../window-manager');
+
+class TrayIconHandler {
+  constructor(appIcon = new Tray(nativeImage.createEmpty())) {
+    this.appIcon = appIcon;
+    this.icons = this.createIcons();
+    this.lastUnreadCount = 0;
+    this.initTrayIcon();
+  }
+
+  initTrayIcon() {
+    this.appIcon.setImage(this.icons.tray);
+    if (this.hasTrayMenuSupport) {
+      this.buildTrayMenu();
+    }
+  }
+
+  get defaultImageExtension() {
+    return environment.platform.IS_WINDOWS ? 'ico' : 'png';
+  }
+
+  get hasTrayMenuSupport() {
+    return !environment.platform.IS_MAC_OS;
+  }
+
+  get hasOverlaySupport() {
+    return environment.platform.IS_WINDOWS;
+  }
+
+  buildTrayMenu() {
+    const contextMenu = Menu.buildFromTemplate([
+      {
+        click: () => windowManager.showPrimaryWindow(),
+        label: locale.getText('trayOpen'),
+      },
+      {
+        click: async () => await lifecycle.quit(),
+        label: locale.getText('trayQuit'),
+      },
+    ]);
+
+    this.appIcon.on('click', () => windowManager.showPrimaryWindow());
+    this.appIcon.setContextMenu(contextMenu);
+    this.appIcon.setToolTip(config.NAME);
+  }
+
+  createIcons() {
+    return {
+      badge: nativeImage.createFromPath(path.join(app.getAppPath(), 'img', 'taskbar.overlay.png')),
+      tray: nativeImage.createFromPath(path.join(app.getAppPath(), 'img', `tray.${this.defaultImageExtension}`)),
+      trayWithBadge: nativeImage.createFromPath(
+        path.join(app.getAppPath(), 'img', `tray.badge.${this.defaultImageExtension}`)
+      ),
+    };
+  }
+
+  updateBadgeIcon(win, count) {
+    if (count) {
+      this.appIcon.setImage(this.icons.trayWithBadge);
+    } else {
+      this.appIcon.setImage(this.icons.tray);
+    }
+
+    if (this.hasOverlaySupport) {
+      win.setOverlayIcon(count ? this.icons.badge : null, locale.getText('unreadMessages'));
+    }
+
+    if (win.isFocused() || !count) {
```

Comment:
how about `win.flashFrame(!win.isFocused() && count && count > this.lastUnreadCount)`

(the condition could be extracted in a variable)

Reply:
I think this condition is becoming very complicated that way. We had a lot of issues with this and @maximbaz fixed it before.

Reply:
Also I believe it won't be equal, it is crucial that we don't call flashFrame in some cases

Reply:
Not sure I get your point guys? `flashFrame` is called either way (falsy or truthy cases)

If you find the condition to hard to read, it could be decomposed into two 

```
const hasFocus = win.isFocused();
const hasNewMessages = count && count > this.lastUnreadCount;

win.flashFrame(hasFocus && hasNewMessages)
```

Reply:
Ohhh ok, I re-read the `else if`, indeed there are some cases when the method is not called.
Since I don't know the consequences of that method, maybe it's better to leave it like this then

Reply:
If the window is not focused, count is 2 and lastUnreadCount is 5, we dont want to call flashFrame at all

Reply:
@maximbaz Thanks for posting this scenario. I will create a test case from it. 😎